src/nvapi/nvapi_d3d12_device.cpp

```diff
@@ -2,6 +2,70 @@
 #include "../util/util_log.h"
 
 namespace dxvk {
+    std::unordered_map<IUnknown*, NvapiD3d12Device> NvapiD3d12Device::m_nvapiDeviceMap;
+    std::mutex NvapiD3d12Device::m_mutex;
+
+    std::unordered_map<NVDX_ObjectHandle, NvU32> NvapiD3d12Device::m_cubinSmemMap;
+    std::mutex NvapiD3d12Device::m_cubinSmemMutex;
+
+    void NvapiD3d12Device::Reset() {
+        std::scoped_lock lock{m_mutex};
+        m_nvapiDeviceMap.clear();
+        m_cubinSmemMap.clear();
+    }
+
+    NvapiD3d12Device* NvapiD3d12Device::GetOrCreate(IUnknown* unknown) {
+        std::scoped_lock lock{m_mutex};
+
+        if (auto nvapiD3D12Device = Get(unknown))
+            return nvapiD3D12Device;
+
+        Com<ID3D12DeviceExt> deviceExt;
+        if (SUCCEEDED(unknown->QueryInterface(IID_PPV_ARGS(&deviceExt))))
+            return Create(unknown, deviceExt.ptr());
+
+        Com<ID3D12GraphicsCommandListExt> commandListExt;
+        if (SUCCEEDED(unknown->QueryInterface(IID_PPV_ARGS(&commandListExt))))
+            return Create(unknown, commandListExt.ptr());
+
+        Com<ID3D12CommandQueueExt> commandQueueExt;
+        if (SUCCEEDED(unknown->QueryInterface(IID_PPV_ARGS(&commandQueueExt))))
+            return Create(unknown, commandQueueExt.ptr());
+
+        return nullptr;
+    }
+
+    template <typename T>
+    NvapiD3d12Device* NvapiD3d12Device::Create(IUnknown* unknown, T* vkd3dUnknown) {
+        auto [it, inserted] = m_nvapiDeviceMap.emplace(
+            std::piecewise_construct,
+            std::forward_as_tuple(unknown),
+            std::forward_as_tuple(vkd3dUnknown));
+
+        if (!inserted)
+            return nullptr;
+
+        return &it->second;
+    }
+
+    NvapiD3d12Device* NvapiD3d12Device::Get(IUnknown* unknown) {
+        auto it = m_nvapiDeviceMap.find(unknown);
+        return it == m_nvapiDeviceMap.end() ? nullptr : &it->second;
+    }
+
+    NvapiD3d12Device::NvapiD3d12Device(ID3D12DeviceExt* vkd3dDevice)
+        : m_vkd3dDevice(vkd3dDevice) {
+        m_supportsNvxBinaryImport = vkd3dDevice->GetExtensionSupport(D3D12_VK_NVX_BINARY_IMPORT);
+    }
+
+    NvapiD3d12Device::NvapiD3d12Device(ID3D12CommandQueueExt* vkd3dCommandQueue)
+        : m_vkd3dCommandQueue(vkd3dCommandQueue) {}
+
+    NvapiD3d12Device::NvapiD3d12Device(ID3D12GraphicsCommandListExt* vkd3dCommandList)
+        : m_vkd3dGraphicsCommandList(static_cast<ID3D12GraphicsCommandListExt1*>(vkd3dCommandList)) { // NOLINT(*-pro-type-static-cast-downcast)
+        Com<ID3D12GraphicsCommandListExt1> commandListExt1 = nullptr;
+        m_supportsExtGraphicsCommandList1 = SUCCEEDED(vkd3dCommandList->QueryInterface(IID_PPV_ARGS(&commandListExt1)));
+    }
```

Comment:
I don't like this. This will make the `NvapiD3d12Device` into a class that's actually one of three things under the hood: a device, a command queue, or a command list. Can we please split this into separate classes? `NvapiD3d12Device`, `NvapiD3d12CommandQueue` and `NvapiD3d12CommandList`, then keep the responsibilities of each class limited to their own.

At worst, the command list one would need an additional method to retrieve smem size from the map technically owned by devices but that shouldn't be too much of a trouble.

(I'm not sure if the map with smem sizes should be a global one or per device but it probably doesn't matter. The former is likely simpler, and it's not like we will have applications running cubin stuff on multiple logical devices at the same time.)